### PR TITLE
Bagatur/refac faiss

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/faiss.py [hunk 1]
```diff
@@ -480,22 +489,26 @@ def delete(self, ids: Optional[List[str]] = None, **kwargs: Any) -> Optional[boo
...
-
-        # Remove items from docstore.
         self.docstore.delete(ids)
+
+        remaining_ids = [
+            id_
+            for i, id_ in sorted(self.index_to_docstore_id.items())
+            if i not in index_to_delete
+        ]
+        self.index_to_docstore_id = {i: id_ for i, id_ in enumerate(remaining_ids)}
```
Comment (eyurtsev) [inline libs/langchain/langchain/vectorstores/faiss.py:510]: Do we have a race condition here that we need to worry about?

File: libs/langchain/langchain/vectorstores/faiss.py [hunk 2]
```diff
@@ -480,22 +489,26 @@ def delete(self, ids: Optional[List[str]] = None, **kwargs: Any) -> Optional[boo
...
         self.index.remove_ids(np.array(index_to_delete, dtype=np.int64))
-        for _id in index_to_delete:
-            del self.index_to_docstore_id[_id]
-
-        # Remove items from docstore.
         self.docstore.delete(ids)
+
+        remaining_ids = [
+            id_
+            for i, id_ in sorted(self.index_to_docstore_id.items())
```
Comment (eyurtsev) [inline libs/langchain/langchain/vectorstores/faiss.py:507]: i and id_ are very closely named

File: libs/langchain/langchain/vectorstores/faiss.py [hunk 3]
```diff
@@ -87,44 +109,43 @@ def __init__(
...
     def __add(
         self,
         texts: Iterable[str],
         embeddings: Iterable[List[float]],
-        metadatas: Optional[List[dict]] = None,
+        metadatas: Optional[Iterable[dict]] = None,
         ids: Optional[List[str]] = None,
-        **kwargs: Any,
     ) -> List[str]:
+        faiss = dependable_faiss_import()
```
[the remark this answers is not in the document]
Comment (baskaryan) [reply libs/langchain/langchain/vectorstores/faiss.py:119]: not really sure, predates me https://github.com/langchain-ai/langchain/pull/119/files#diff-835ff9ac82717f240be146d5d7b3f8922064d5756f34aba34ba94f72f1658807

File: libs/langchain/langchain/vectorstores/faiss.py (hunk 1 above)
Comment (baskaryan) [reply libs/langchain/langchain/vectorstores/faiss.py:510]: like if the docstore or index haven't finished deleting?

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**